R/process_observers.R

```diff
@@ -0,0 +1,170 @@
+      #                     primary_observer = replace(primary_observer, 
+      #                                                county == selected_county & #match county
+      #                                                route == rocombos$route_num[i] & #& route
+      #                                                observers == edit_row, #& observers
+      #                                                new_primaryobs)) #replace w/new primaryobs
+      #   } else {print("Cannot find observers as typed, moving to next new route/observer combo" )}
+      
+    } #end else statement about this observer/route combo not being on the conversion table
+  }  #end for loop (done going through all the rocombos)
+  
+  #save updated version of observer conversion table
+  save_observer_table(observer_table)
+  print("No more new route/observer combos. Done!")
+} #end function
+
+
+
+#' Pull out observers from the checklist_comments column that process_comments missed
+#' @param mbbs_county mbbs data.frame
+#' @importFrom stringr str_replace_all str_detect
+#' @importFrom dplyr mutate case_when
+#' @export
+observers_extractor <- function(mbbs_county) {
+  
+  #fix unicode 
+  mbbs_county$checklist_comments <- mbbs_county$checklist_comments %>%
+    str_replace_all(c("&#61;"), "=")
+  
+  #when checklist comments contain "observer(s)", extract after observer(s) and before a ;
+  mbbs_county <- mbbs_county %>% mutate(
+    observers = case_when(
+      is.na(observers) == FALSE ~ observers, #if observer column exists leave it
+      #if observer column is NA, extract from comments after observer and before ;
+      str_detect(checklist_comments,".*[oO]bserver(s)?=") == TRUE ~
+        sub(".*[oO]bserver(s)?=", "", checklist_comments) %>% #extract after observer
+        {sub(";.*", "", .)}, #extract before ;
+      #if observer column is NA but comments doesn't include observers, leave as is
+      str_detect(checklist_comments,".*[oO]bserver(s)?=") == FALSE ~ observers
+    ))
+  
+  return(mbbs_county)
+}
+
+
+
+#' Function to add observers to route stops 2:20
+#' @param mbbs_county mbbs data.frame
+#' @importFrom dplyr group_by mutate
+#' @export
+propogate_observers_across_stops <- function(mbbs_county) {
+  #group by route and date (all observations on that route, inclusive of stops 1:20 on ebird checklists after 2019) and give all observer columns the same value as whatever column is not NA
+  mbbs_county <-  mbbs_county %>% group_by(route_num, date) %>% mutate(observers = max(observers, na.rm  = T))
+  #will fill in stops 2:20 with checklist comments like v;3 and won't change data from pre-2019 because all the observations on the same route_num and date will already have the same comments/observer columns
+  return(mbbs_county)
+}
+
+
+
+
+#' Full workflow for processing observers 
+#' @param mbbs_county mbbs data.frame
+#' @importFrom dplyr
+#' @export
+process_observers(mbbs_county) {
+  mbbs_county <- mbbs_county %>% 
```

Comment:
This can be simplified to:

```r
  mbbs_county %>% 
    observers_extractor() %>% 
    propogate_observers_across_stops() %>%
    update_observer_table()


```